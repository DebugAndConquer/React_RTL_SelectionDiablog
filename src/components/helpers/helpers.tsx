import { ISelectorWidgetOption } from "../../types/SelectorWidgetTypes";

// Immutably updates an isSelected attribute for the entry with provided value
export const setValuesWrapper = (
  value: number,
  isSelected: boolean,
  setter: (value: React.SetStateAction<ISelectorWidgetOption[]>) => void
): void => {
  setter((prev) => {
    return prev.map((el) => {
      // When the value matches the searched one -> update isSelected attribute
      if (el.value === value) {
        return {
          ...el,
          isSelected: isSelected,
        };
        // When no match -> leave as it is
      } else {
        return el;
      }
    });
  });
};

// Generates the specified number of list elements with name 'Element <x>'
export const generateSampleData = (
  numberOfEl: number
): ISelectorWidgetOption[] => {
  const sampleData: ISelectorWidgetOption[] = [];
  for (let i = 1; i <= numberOfEl; i++) {
    sampleData.push({ label: `Element ${i}`, value: i, isSelected: false });
  }
  return sampleData;
};

export const spacesToDash = (initialString: string): string =>
  initialString.replaceAll(" ", "-");
