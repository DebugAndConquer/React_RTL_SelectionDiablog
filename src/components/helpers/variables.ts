export const ELEMENT_COUNT = 300; // Total test elements to create

export const ACTIVE_ENTRY = "selector-dialog__entries__entry"; // Regular entry class
export const DISABLED_ENTRY_MOD = "selector-dialog__entries__entry--disabled"; // A modifier to make it disabled
export const MAXIMUM_SELECTED_ENTRIES = 3; // No more than specified number of entries will be allowed to be selected

// FILTERS
export const FILTER_VALUES = {
    FILTER_NONE: 0,
    FILTER_MORE_THAN_10: 1,
    FILTER_MORE_THAN_50: 2,
    FILTER_MORE_THAN_100: 3
}