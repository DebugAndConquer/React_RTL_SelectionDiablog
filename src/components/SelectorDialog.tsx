import React, { type FC, useEffect, useState } from 'react'
import { FilterFunctionType, ISelectorWidgetOption } from "../types/SelectorWidgetTypes";
import classNames from "classnames";
import { setValuesWrapper } from './helpers/helpers';
import SelectorOptions from './SelectorOptions';
import { ACTIVE_ENTRY, DISABLED_ENTRY_MOD, FILTER_VALUES, MAXIMUM_SELECTED_ENTRIES } from './helpers/variables';
import Search from './Search';
import Filter from './Filter';

interface ISelectorDialogProps {
    onSave: (newValues: ISelectorWidgetOption[]) => void,
    onHide: () => void,
    options: ISelectorWidgetOption[]
}

const SelectorDialog: FC<ISelectorDialogProps> = ({ options, onSave, onHide }) => {
    const [values, setValues] = useState<ISelectorWidgetOption[]>([])
    const [entryClasses, setEntryClasses] = useState<string>("");
    const [search, setSearch] = useState<string>("");
    const [filterFunction, setFilterFunction] = useState<FilterFunctionType>(() => ((input: ISelectorWidgetOption[]) => input));

    useEffect(() => {
        setValues(options)
    }, [options])

    // Dynamic classnames 
    useEffect(() => {
        setEntryClasses(classNames({
            [ACTIVE_ENTRY]: true,
            [DISABLED_ENTRY_MOD]: values.filter((x) => x.isSelected).length >= MAXIMUM_SELECTED_ENTRIES
        }))
    }, [values])

    const onSelect = (opt: ISelectorWidgetOption): void => {
        if (opt.isSelected === true) {
            setValuesWrapper(opt.value, false, setValues);
        } else if (opt.isSelected === false) {
            setValuesWrapper(opt.value, true, setValues);
        }
    }

    const handleSearch = (e: React.ChangeEvent<HTMLInputElement>): void => {
        setSearch(e.target.value.toLowerCase());
    }

    const handleFilterChange = (e: React.ChangeEvent<HTMLSelectElement>): void => {
        const valueSelected = e.target.value
        switch (parseInt(valueSelected)) {
            case FILTER_VALUES.FILTER_MORE_THAN_10:
                setFilterFunction(() => ((input: ISelectorWidgetOption[]) => input.filter((v) => v.value > 10)));
                break;
            case FILTER_VALUES.FILTER_MORE_THAN_50:
                setFilterFunction(() => ((input: ISelectorWidgetOption[]) => input.filter((v) => v.value > 50)));
                break;
            case FILTER_VALUES.FILTER_MORE_THAN_100:
                setFilterFunction(() => ((input: ISelectorWidgetOption[]) => input.filter((v) => v.value > 100)));
                break;
            case FILTER_VALUES.FILTER_NONE:
            default:
                setFilterFunction(() => ((input: ISelectorWidgetOption[]) => input));
                break;

        }
    }
    const filteredValues = filterFunction(values).filter((v) => v.label.toLowerCase().includes(search));
    return (
        <div className='selector-dialog'>
            <p>Element Selection Dialog</p>
            <div className="selector-dialog__filters">
                <Search onChange={handleSearch} testId="mainSearchElement"/>
                <Filter onChange={handleFilterChange} testId="mainFilterElement" />
            </div>
            <div className="selector-dialog__entries">
                {values && filteredValues.map((opt) => (
                    <div key={opt.value} onClick={() => onSelect(opt)} className={opt.isSelected ? ACTIVE_ENTRY : entryClasses}>
                        <input type={"checkbox"} readOnly checked={opt.isSelected} />
                        <p>{opt.label}</p>
                    </div>
                ))}
                {
                    values.length > 0 && filteredValues.length === 0 && <p>Nothing matches your query</p>
                }

            </div>
            <p>Currently selected elements:</p>
            <SelectorOptions values={values} handleUnselect={(value) => setValuesWrapper(value, false, setValues)} />
            <div className='selector-dialog__btns'>
                <button className='btn btn--green' onClick={() => onSave(values)}>Save</button>
                <button className='btn btn--red' onClick={() => onHide()}>Cancel</button>
            </div>
        </div>
    )
}

export default SelectorDialog