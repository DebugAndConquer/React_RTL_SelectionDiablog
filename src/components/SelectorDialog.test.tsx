import React from "react";
import { prettyDOM, render, screen, waitFor } from "@testing-library/react";
import { generateSampleData } from "./helpers/helpers";
import userEvent from "@testing-library/user-event";
import SelectorDialog from "./SelectorDialog";

const onSave = jest.fn();
const onHide = jest.fn();

test("Matches the snapshot", () => {
  const { container } = render(
    <SelectorDialog
      options={generateSampleData(10)}
      onSave={onSave}
      onHide={onHide}
    />
  );
  expect(container).toMatchSnapshot();
});

test("Verifies that search filters the list correctly", async () => {
  render(
    <SelectorDialog
      // Generate a list of 10 elements for this test
      options={generateSampleData(10)}
      onSave={onSave}
      onHide={onHide}
    />
  );
  const searchElement = screen.getByTestId("mainSearchElement");
  userEvent.type(searchElement, "Element 5");
  expect(screen.getByText("Element 5")).toBeInTheDocument();
  expect(screen.queryByText("Element 4")).not.toBeInTheDocument();
  await userEvent.clear(searchElement);

  userEvent.type(searchElement, "1");
  expect(screen.getByText("Element 1")).toBeInTheDocument();
  expect(screen.getByText("Element 10")).toBeInTheDocument();
  expect(screen.queryByText("Element 4")).not.toBeInTheDocument();
  expect(screen.queryByText("Element 11")).not.toBeInTheDocument();
});

test("Verifies that filter works as expected and changes the list accordingly", () => {
  render(
    <SelectorDialog
      // Generate a list of 300 elements for this test
      options={generateSampleData(300)}
      onSave={onSave}
      onHide={onHide}
    />
  );
  const filterElement = screen.getByTestId("mainFilterElement");
  userEvent.selectOptions(filterElement, "Number &gt; 10");
  expect(screen.queryByText("Element 4")).not.toBeInTheDocument();
  expect(screen.queryByText("Element 10")).not.toBeInTheDocument();
  expect(screen.getByText("Element 11")).toBeInTheDocument();
  expect(screen.getByText("Element 200")).toBeInTheDocument();

  userEvent.selectOptions(filterElement, "Number &gt; 50");
  expect(screen.queryByText("Element 4")).not.toBeInTheDocument();
  expect(screen.queryByText("Element 10")).not.toBeInTheDocument();
  expect(screen.queryByText("Element 50")).not.toBeInTheDocument();
  expect(screen.getByText("Element 51")).toBeInTheDocument();
  expect(screen.getByText("Element 200")).toBeInTheDocument();

  userEvent.selectOptions(filterElement, "Number &gt; 100");
  expect(screen.queryByText("Element 4")).not.toBeInTheDocument();
  expect(screen.queryByText("Element 50")).not.toBeInTheDocument();
  expect(screen.queryByText("Element 100")).not.toBeInTheDocument();
  expect(screen.getByText("Element 101")).toBeInTheDocument();
  expect(screen.getByText("Element 200")).toBeInTheDocument();

  userEvent.selectOptions(filterElement, "No filters");
  expect(screen.getByText("Element 1")).toBeInTheDocument();
  expect(screen.getByText("Element 11")).toBeInTheDocument();
  expect(screen.getByText("Element 201")).toBeInTheDocument();
});

test("Verifies that element selection works, the black card is created for each selected element and elements can be removed", async () => {
  render(
    <SelectorDialog
      options={generateSampleData(10)}
      onSave={onSave}
      onHide={onHide}
    />
  );
  // Element selection assertations
  const element2 = screen.getByText("Element 2");
  const element5 = screen.getByText("Element 5");
  await userEvent.click(element2);
  expect(screen.getAllByText("Element 2")).toHaveLength(2);
  expect(screen.getByTestId("element-2-selector-option")).toBeInTheDocument();
  expect(screen.getAllByText("Element 5")).toHaveLength(1);
  expect(
    screen.queryByTestId("element-5-selector-option")
  ).not.toBeInTheDocument();

  await userEvent.click(element5);
  expect(screen.getAllByText("Element 2")).toHaveLength(2);
  expect(screen.getByTestId("element-2-selector-option")).toBeInTheDocument();
  expect(screen.getAllByText("Element 5")).toHaveLength(2);
  expect(screen.getByTestId("element-5-selector-option")).toBeInTheDocument();

  // Element de-selection assertations
  // By clicking again on the element
  await userEvent.click(element2);
  expect(screen.getAllByText("Element 2")).toHaveLength(1);
  expect(
    screen.queryByTestId("element-2-selector-option")
  ).not.toBeInTheDocument();
  // By clicking on black box's close button
  await userEvent.click(
    screen.getByTestId("element-5-selector-option-remove-button")
  );
  expect(screen.getAllByText("Element 5")).toHaveLength(1);
  expect(
    screen.queryByTestId("element-5-selector-option")
  ).not.toBeInTheDocument();
  expect(
    screen.getByText(/Selected elements will appear here/i)
  ).toBeInTheDocument();
});

test("Verifies that save and cancel buttons work and handlers are being called", async () => {
  render(
    <SelectorDialog
      options={generateSampleData(3)}
      onSave={onSave}
      onHide={onHide}
    />
  );

  const element1 = screen.getByText("Element 1");
  const element2 = screen.getByText("Element 2");
  const element3 = screen.getByText("Element 3");
  const saveButtonElement = screen.getByText(/Save/i);
  const cancelButtonElement = screen.getByText(/Cancel/i);

  await userEvent.click(element1);
  await userEvent.click(saveButtonElement);
  expect(onSave).toHaveBeenCalledWith([
    { isSelected: true, label: "Element 1", value: 1 },
    { isSelected: false, label: "Element 2", value: 2 },
    { isSelected: false, label: "Element 3", value: 3 },
  ]);

  await userEvent.click(element2);
  await userEvent.click(saveButtonElement);
  expect(onSave).toHaveBeenCalledWith([
    { isSelected: true, label: "Element 1", value: 1 },
    { isSelected: true, label: "Element 2", value: 2 },
    { isSelected: false, label: "Element 3", value: 3 },
  ]);

  await userEvent.click(element3);
  await userEvent.click(cancelButtonElement);
  expect(onSave).toHaveBeenCalledWith([
    { isSelected: true, label: "Element 1", value: 1 },
    { isSelected: true, label: "Element 2", value: 2 },
    { isSelected: false, label: "Element 3", value: 3 },
  ]);
  expect(onHide).toHaveBeenCalled();
});
