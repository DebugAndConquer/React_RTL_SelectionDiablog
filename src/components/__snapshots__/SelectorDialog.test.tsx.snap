// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches the snapshot 1`] = `
<div>
  <div
    class="selector-dialog"
  >
    <p>
      Element Selection Dialog
    </p>
    <div
      class="selector-dialog__filters"
    >
      <div>
        <label>
          Search
        </label>
        <input
          data-testid="mainSearchElement"
          placeholder="Start typing..."
          type="search"
        />
      </div>
      <div>
        <label>
          Filter
        </label>
        <select
          data-testid="mainFilterElement"
        >
          <option
            selected=""
            value="0"
          >
            No filters
          </option>
          <option
            value="1"
          >
            Number &gt; 10
          </option>
          <option
            value="2"
          >
            Number &gt; 50
          </option>
          <option
            value="3"
          >
            Number &gt; 100
          </option>
        </select>
      </div>
    </div>
    <div
      class="selector-dialog__entries"
    >
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 1
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 2
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 3
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 4
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 5
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 6
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 7
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 8
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 9
        </p>
      </div>
      <div
        class="selector-dialog__entries__entry"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <p>
          Element 10
        </p>
      </div>
    </div>
    <p>
      Currently selected elements:
    </p>
    <div
      class="sel-widget__selected-opt-cont"
    >
      <p
        class="no-selected-options"
      >
        Selected elements will appear here
      </p>
    </div>
    <div
      class="selector-dialog__btns"
    >
      <button
        class="btn btn--green"
      >
        Save
      </button>
      <button
        class="btn btn--red"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
